### PR TITLE
Native skia bindings

### PR DESCRIPTION
**Pull request by alexander-gorshenev**

### CODE REVIEW

**Comment by kirill-grouchnikov** [comment]
https://github.com/JetBrains/skiko/pull/165/files has added two functions in shader.cc - they are missing in the generated skiko/src/macosX64Main/cpp/generated/Shader.cc

File: skiko/src/commonMain/cpp/common/common_interop.cc [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+jlong packISize(SkISize p) {
+    return packTwoInts(p.fWidth, p.fHeight);
+}
+
+namespace skija {
+
+    namespace SamplingMode {
+        SkSamplingOptions unpack(jlong val) {
+            if (0x8000000000000000 & val) {
+                val = val & 0x7FFFFFFFFFFFFFFF;
```
Comment (olonho) [inline skiko/src/commonMain/cpp/common/common_interop.cc:21]: Better use union to avoid aliasing issues. Also not sure if I understand the logic here, we use upper bit to select between resamplers?

Comment (alexander-gorshenev) [reply skiko/src/commonMain/cpp/common/common_interop.cc:21]: This code has just been moved verbatim 
from `src/jvmMain/cpp/common/interop.cc` 
to `src/commonMain/cpp/common/common_interop.cc `.
The jvm jni code still depends on it.
I'd like to make whatever semantic changes as a separate fix in the bindings-- this fix is HUGE already.

File: samples/SkiaNativeSample/src/macosX64Main/kotlin/org.jetbrains.skiko.sample.native/App.kt [hunk 2]
```diff
@@ -50,14 +51,17 @@ fun displayScene(renderer: Renderer, nanoTime: Long) {
     val canvas = renderer.canvas!!
 
     val paint = Paint()
-    paint.setColor(SK_ColorGREEN)
+    //paint.setColor(Color.GREEN)
+    paint.setColor(0xFF00FF00.toInt())
```
Comment (olonho) [inline samples/SkiaNativeSample/src/macosX64Main/kotlin/org.jetbrains.skiko.sample.native/App.kt:55]: Guess we shall instead create appropriate constants in Color object.

File: samples/SkiaNativeSample/src/macosX64Main/kotlin/org.jetbrains.skiko.sample.native/App.kt [hunk 3]
```diff
@@ -50,14 +51,17 @@ fun displayScene(renderer: Renderer, nanoTime: Long) {
...
 
-    canvas.clear(SK_ColorRED);
+    // canvas.clear(Color.RED);
+    canvas.clear(0x00FF00FF.toInt());
 
     canvas.save();
-    canvas.translate(128.0f, 128.0f)
-    canvas.rotate(nanoTime.toFloat() / 1e7f)
-    val rect = SkRect.MakeXYWH(-90.5f, -90.5f, 181.0f, 181.0f)
+    // TODO: disabled to ramp up the new native skiko.
```
Comment (olonho) [inline samples/SkiaNativeSample/src/macosX64Main/kotlin/org.jetbrains.skiko.sample.native/App.kt:61]: What's wrong here, why we shall do that?

File: skiko/build.gradle.kts [hunk 4]
```diff
@@ -127,22 +126,29 @@ kotlin {
...
+                        freeCompilerArgs += listOf(
+                            "-include-binary",
+                            "$skiaDir/$skiaBinSubdir/libskia.a",
+                            "-include-binary",
+                            "$skiaDir/$skiaBinSubdir/libskshaper.a",
+                            "-include-binary",
+                            "$skiaDir/$skiaBinSubdir/libskparagraph.a"
+                        )
+
+                        freeCompilerArgs += listOf(
```
Comment (olonho) [inline skiko/build.gradle.kts:144]: Why not join with above?

File: skiko/build.gradle.kts [hunk 5]
```diff
@@ -181,32 +187,19 @@ kotlin {
 
         if (supportNative) {
             val macosX64Main by getting {
+                dependsOn(commonMain)
```
Comment (olonho) [inline skiko/build.gradle.kts:190]: Is it actually required?

File: skiko/build.gradle.kts [hunk 6]
```diff
@@ -354,6 +368,75 @@ project.tasks.register<Exec>("wasmCompile") {
...
+        .map { it.substringAfterLast("/") }
+        .map { File(it).nameWithoutExtension }
+        .map { "$outDir/$it.o" }
+        .toTypedArray()
+
+    workingDir = File(outDir)
+    commandLine = listOf(
+        "clang++",
+        *targetArch.clangFlags,
+        "-mmacosx-version-min=10.13",
```
Comment (olonho) [inline skiko/build.gradle.kts:401]: Better move that to argetArch.clangFlags

File: skiko/build.gradle.kts [hunk 7]
```diff
@@ -354,6 +368,75 @@ project.tasks.register<Exec>("wasmCompile") {
...
+
+    val outDir = "$buildDir/nativeBridges/static/$target"
+    val srcs = inputs.files.files
+        .map { it.absolutePath }
+        .toTypedArray()
+    println("SRCS: $srcs")
+    val staticLib = "$outDir/skiko-native-bridges-$target.a"
+    workingDir = File(outDir)
+    commandLine = listOf(
+        "libtool",
```
Comment (olonho) [inline skiko/build.gradle.kts:430]: Why not just ar directly?

File: skiko/build.gradle.kts [hunk 8]
```diff
@@ -354,6 +368,75 @@ project.tasks.register<Exec>("wasmCompile") {
...
+        "clang++",
+        *targetArch.clangFlags,
+        "-mmacosx-version-min=10.13",
+        "-std=c++17",
+        "-c",
+        "-DPROVIDE_JNI_TYPES",
+        "-DSK_SHAPER_CORETEXT_AVAILABLE",
+        "-DSK_BUILD_FOR_MAC",
+        "-DSK_METAL",
+        "-I$projectDir/src/macosX64Main/cpp/headers",
```
Comment (olonho) [inline skiko/build.gradle.kts:408]: Can we rename macosX64Main -> nativeMain and depend in macosX64Main on it. Can do that in later PRs

File: skiko/src/commonMain/cpp/common/common_interop.cc (hunk 1 above)
Comment (olonho) [reply skiko/src/commonMain/cpp/common/common_interop.cc:21]: OK

File: skiko/src/commonMain/cpp/headers/common.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+#include <stdexcept>
+#include "types.h"
+
+jlong packTwoInts(int32_t a, int32_t b);
+
+jlong packIPoint(SkIPoint p);
+
+jlong packISize(SkISize p);
+
+namespace skija {
```
Comment (olonho) [inline skiko/src/commonMain/cpp/headers/common.h:27]: Probably we shall rename skija -> skiko eventually.

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/Canvas.kt [hunk 10]
```diff
@@ -1298,11 +1338,13 @@ open class Canvas internal constructor(ptr: Long, managed: Boolean, internal val
     }
 
     fun translate(dx: Float, dy: Float): Canvas {
-        return concat(Matrix33.makeTranslate(dx, dy))
+        val matrix = Matrix33.makeTranslate(dx, dy)
```
Comment (olonho) [inline skiko/src/commonMain/kotlin/org/jetbrains/skia/Canvas.kt:1341]: Why this change?

File: skiko/build.gradle.kts (hunk 7 above)
Comment (alexander-gorshenev) [reply skiko/build.gradle.kts:430]: However I tried, all binaries produced with `ar` are ignored by linker. The binaries produced with` libtool` just work.

File: skiko/build.gradle.kts (hunk 8 above)
Comment (alexander-gorshenev) [reply skiko/build.gradle.kts:408]: I have that in my close plans, yes

File: skiko/src/commonMain/cpp/headers/common.h (hunk 9 above)
Comment (alexander-gorshenev) [reply skiko/src/commonMain/cpp/headers/common.h:27]: yes, I suppose.

File: samples/SkiaNativeSample/src/macosX64Main/kotlin/org.jetbrains.skiko.sample.native/App.kt (hunk 3 above)
Comment (alexander-gorshenev) [reply samples/SkiaNativeSample/src/macosX64Main/kotlin/org.jetbrains.skiko.sample.native/App.kt:61]: Those eventually pass a Matrix, which goes to the native bindings as `jfloatArray` or something. Arrays are not working yet.

File: skiko/build.gradle.kts (hunk 4 above)
Comment (alexander-gorshenev) [reply skiko/build.gradle.kts]: joined

File: skiko/build.gradle.kts (hunk 5 above)
Comment (alexander-gorshenev) [reply skiko/build.gradle.kts]: dropped

File: skiko/build.gradle.kts (hunk 6 above)
Comment (alexander-gorshenev) [reply skiko/build.gradle.kts]: moved to targetOs.clangFlags

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/Canvas.kt (hunk 10 above)
Comment (alexander-gorshenev) [reply skiko/src/commonMain/kotlin/org/jetbrains/skia/Canvas.kt]: not needed, dropped

File: skiko/build.gradle.kts (hunk 7 above)
Comment (olonho) [reply skiko/build.gradle.kts:426]: Ah, maybe you need to run ranlib, but anyway - it's simpler this way then.

**Review by olonho — APPROVED** [review]
